README.md
# shopify-api v0.2.3 [![Build Status](https://secure.travis-ci.org/ctalkington/node-shopify-api.png?branch=master)](http://travis-ci.org/ctalkington/node-shopify-api)

shopify-api is a node library for working with the [Shopify API](http://docs.shopify.com/api). It is designed to abstract away requests while returning unmodified responses. It should be used as a building block for other modules/packages which can handle any additional abstraction needed based on the API resource that they are working with.

### Install

```bash
npm install shopify-api --save
```

You can also use `npm install https://github.com/ctalkington/node-shopify-api/archive/master.tar.gz` to test upcoming versions.

### Usage

This module is meant to be wrapped internally by other modules and therefore lacks any frontend interface.

Every resource is accessed via your `shopify` instance:

```js
var shopify = require('shopify-api')(options);
// shopify.{ RESOURCE_NAME }.{ METHOD_NAME }
```

Every resource method accepts an optional callback as the last argument:

```js
shopify.asset.create(
  { key: '...', value: '...' },
  function(err, asset) {
    if (err) {
      throw err;
    }
    console.log(asset);
  }
);
```

Additionally, every resource method returns a promise, so you don't have to use the regular callback. E.g.

```js
shopify.asset.create({
  key: '...',
  value: '...'
}).then(function(customer) {
  // do something
}, function(err) {
  // Deal with an error
});
```
### Instance Options

#### auth

The auth for requests to use.

*This must be set either via option or setter method.*

#### host

The host for requests to use.

*This must be set either via option or setter method.*

#### port

The port for requests to use. Default: 443

#### timeout

The time in `ms` to wait for requests before timing out. Default: 120000ms (2 minutes)

### Instance Methods

#### setAuth(key, passwd)

Sets the internal `auth` property.

#### setHost(host[, port])

Sets the internal `host`/`port` properties.

#### setPort(port)

Sets the internal `port` property.

#### setTimeout(msec)

Sets the internal `timeout` property.

### Resources

The resources that are currently supported by this library. Please submit PR if your need others.

#### asset

##### create(themeid, props)

##### destroy(themeid, key)

##### list(themeid)

##### retrieve(themeid, key)

##### update(themeid, props)

#### assetLegacy

##### create(props)

##### destroy(key)

##### list()

##### retrieve(key)

##### update(props)

#### theme

##### create(props)

##### destroy(id)

##### list()

##### retrieve(id)

##### update(id, props)

## Things of Interest

- [Changelog](https://github.com/ctalkington/node-shopify-api/releases)
- [Contributing](https://github.com/ctalkington/node-shopify-api/blob/master/CONTRIBUTING.md)
- [MIT License](https://github.com/ctalkington/node-shopify-api/blob/master/LICENSE-MIT)

## Credits

Structured after the [stripe](https://npmjs.org/package/stripe) package.

CHANGELOG
v0.2.3:
  date: 2014-02-24
  changes:
    - catch HTTP 404 errors.
    - add fallback values for error properties.
v0.2.2:
  date: 2014-02-24
  changes:
    - fix error type def.
v0.2.1:
  date: 2014-02-23
  changes:
    - refactor error handling.
    - catch api call limit status code and parse it for other libraries to use.
v0.2.0:
  date: 2014-02-20
  changes:
    - add assetLegacy which doesn't require a themeid but isn't documented all that well.
v0.1.0:
  date: 2014-01-15
  changes:
    - initial version.

test/util.js
/*global before,describe,it */
var assert = require('chai').assert;

describe('util', function() {

  it.skip('should do something at some point', function() {

  });

});

test/shopify.js
/*global before,describe,it */
var assert = require('chai').assert;
var mkdir = require('mkdirp');
var nock = require('nock');

var shopify = require('../lib/shopify');

var instance;
var mockApi;

var themeId = 828155753;

describe('shopify', function() {

  before(function() {
    mkdir.sync('tmp');

    mockApi = nock('https://test.myshopify.com');
    instance = shopify({
      host: 'test.myshopify.com',
      auth: 'key:passwd'
    });
  });

  describe('general', function() {
    it('should catch api errors', function(done) {
      mockApi.get('/admin/themes.json').reply(500);

      instance.theme.list(function(err, data) {
        assert.propertyVal(err, 'type', 'ShopifyAPIError');
        assert.isNull(data);
        done();
      });
    });

    it('should catch auth failures', function(done) {
      mockApi.get('/admin/themes.json').reply(401);

      instance.theme.list(function(err, data) {
        assert.propertyVal(err, 'type', 'ShopifyAuthenticationError');
        assert.isNull(data);
        done();
      });
    });

    it('should catch and parse api call limits', function(done) {
      mockApi.get('/admin/themes.json').reply(429, '', {
        'X-Shopify-Shop-Api-Call-Limit': '50/50'
      });

      instance.theme.list(function(err, data) {
        assert.propertyVal(err, 'type', 'ShopifyCallLimitError');
        assert.deepProperty(err, 'detail.called');
        assert.deepProperty(err, 'detail.limit');
        assert.isNull(data);
        done();
      });
    });

    it('should catch errors passed via JSON', function(done) {
      mockApi.put('/admin/themes/828155753/assets.json').reply(200, {
        "errors": [{
          "key": "is already taken",
        }]
      });

      var props = {
        key: 'assets/empty.gif',
        attachment: "R0lGODlhAQABAPABAP///wAAACH5BAEKAAAALAAAAAABAAEAAAICRAEAOw==\n"
      };

      instance.asset.create(themeId, props, function(err, data) {
        assert.propertyVal(err, 'type', 'ShopifyInvalidRequestError');
        assert.property(err, 'detail');
        assert.isArray(err.detail);
        assert.isNull(data);
        done();
      });
    });

  });

  describe('asset', function() {

    it('should retrieve a list of assets', function(done) {
      mockApi.get('/admin/themes/828155753/assets.json').reply(200, {
        "assets": [{
          "key": "assets/bg-body-green.gif",
          "public_url": "http://cdn.shopify.com/s/files/1/0006/9093/3842/t/1/assets/bg-body-green.gif?1",
          "created_at": "2010-07-12T15:31:50-04:00",
          "updated_at": "2010-07-12T15:31:50-04:00",
          "content_type": "image/gif",
          "size": 1542,
          "theme_id": 828155753
        }]
      });

      instance.asset.list(themeId, function(err, data) {
        assert.isArray(data.assets);
        assert.isObject(data.assets[0]);
        done();
      });
    });

    it('should retrieve a single asset', function(done) {
      mockApi.get('/admin/themes/828155753/assets.json?asset[key]=templates/index.liquid&theme_id=828155753').reply(200, {
        "asset": [{
          "key": "templates/index.liquid",
          "public_url": null,
          "value": "<!-- LIST 3 PER ROW -->\n<h2>Featured Products</h2>\n<table id=\"products\" cellspacing=\"0\" cellpadding=\"0\">\n\n</table>\n<!-- /LIST 3 PER ROW  -->\n\n  <div id=\"articles\">\n  \t\n\n    <div class=\"article\">\n    \n      <div class=\"article-body textile\">\n  \t  In <em>Admin &gt; Blogs &amp; Pages</em>, create a page with the handle <strong><code>frontpage</code></strong> and it will show up here.<br />\n  \t  \n      </div>\n  \t\n    </div>\n\n  </div>\n\n",
          "created_at": "2010-07-12T15:31:50-04:00",
          "updated_at": "2010-07-12T15:31:50-04:00",
          "content_type": "text/x-liquid",
          "size": 1068,
          "theme_id": 828155753
        }]
      });

      instance.asset.retrieve(themeId, 'templates/index.liquid', function(err, data) {
        assert.isArray(data.asset);
        assert.isObject(data.asset[0]);
        done();
      });
    });

    it('should create a new asset', function(done) {
      mockApi.put('/admin/themes/828155753/assets.json').reply(200, {
        "asset": [{
          "key": "assets/empty.gif",
          "public_url": "http://cdn.shopify.com/s/files/1/0006/9093/3842/t/1/assets/empty.gif?2",
          "created_at": "2014-02-21T14:48:38-05:00",
          "updated_at": "2014-02-21T14:48:38-05:00",
          "content_type": "image/gif",
          "size": 43,
          "theme_id": 828155753
        }]
      });

      var props = {
        key: 'assets/empty.gif',
        attachment: "R0lGODlhAQABAPABAP///wAAACH5BAEKAAAALAAAAAABAAEAAAICRAEAOw==\n"
      };

      instance.asset.create(themeId, props, function(err, data) {
        assert.isArray(data.asset);
        assert.isObject(data.asset[0]);
        done();
      });
    });

    it('should update an existing asset', function(done) {
      mockApi.put('/admin/themes/828155753/assets.json').reply(200, {
        "asset": [{
          "key": "templates/index.liquid",
          "public_url": null,
          "created_at": "2010-07-12T15:31:50-04:00",
          "updated_at": "2014-02-21T14:48:38-05:00",
          "content_type": "text/x-liquid",
          "size": 110,
          "theme_id": 828155753
        }]
      });

      var props = {
        key: 'templates/index.liquid',
        value: "<img src='backsoon-postit.png'><p>We are busy updating the store for you and will be back within the hour.</p>"
      };

      instance.asset.update(themeId, props, function(err, data) {
        assert.isArray(data.asset);
        assert.isObject(data.asset[0]);
        done();
      });
    });

  });

  describe('theme', function() {

    it('should retrieve a list of themes', function(done) {
      mockApi.get('/admin/themes.json').reply(200, {
        "themes": [{
          "created_at": "2014-02-21T14:52:02-05:00",
          "id": themeId,
          "name": "Comfort",
          "role": "main",
          "theme_store_id": null,
          "updated_at": "2014-02-21T14:52:02-05:00",
          "previewable": true
        }]
      });

      instance.theme.list(function(err, data) {
        assert.isArray(data.themes);
        assert.isObject(data.themes[0]);
        done();
      });
    });

  });

});

lib/error.js
/**
 * node-shopify-api
 *
 * Copyright (c) 2014 Chris Talkington, contributors.
 * Licensed under the MIT license.
 * https://github.com/ctalkington/node-shopify-api/blob/master/LICENSE-MIT
 */
var util = require('./util');

module.exports = _Error = function(raw) {
  this.populate.apply(this, arguments);
};

// Extend Native Error
_Error.prototype = Object.create(Error.prototype);

_Error.prototype.type = 'GenericError';
_Error.prototype.populate = function(type, message) {
  this.type = type;
  this.message = message;
};

_Error.extend = util.extend;

/**
 * Create subclass of internal Error klass
 */
var ShopifyError = _Error.ShopifyError = _Error.extend({
  type: 'ShopifyError',
  populate: function(raw) {
    this.type = this.type;

    this.message = raw.message || 'Error Missing Message';
    this.detail = raw.detail || false;
    this.raw = raw;
  }
});

_Error.ShopifyInvalidRequestError = ShopifyError.extend({ type: 'ShopifyInvalidRequestError' });
_Error.ShopifyAPIError = ShopifyError.extend({ type: 'ShopifyAPIError' });
_Error.ShopifyAuthenticationError = ShopifyError.extend({ type: 'ShopifyAuthenticationError' });
_Error.ShopifyCallLimitError = ShopifyError.extend({ type: 'ShopifyCallLimitError' });
_Error.ShopifyConnectionError = ShopifyError.extend({ type: 'ShopifyConnectionError' });

lib/method.js
/**
 * node-shopify-api
 *
 * Copyright (c) 2014 Chris Talkington, contributors.
 * Licensed under the MIT license.
 * https://github.com/ctalkington/node-shopify-api/blob/master/LICENSE-MIT
 */
var util = require('./util');

module.exports = function stripeMethod(spec) {
  var methodPath = util.makeInterpolator(spec.path || '');
  var requestMethod = (spec.method || 'GET').toUpperCase();
  var urlParams = spec.urlParams || [];

  return function() {
    var self = this;
    var args = [].slice.call(arguments);

    var callback = util.isFunction(args[args.length - 1]) && args.pop();
    var data = util.isObject(args[args.length - 1]) ? args.pop() : {};
    var urlData = this.createUrlData();

    var deferred = this.createDeferred(callback);

    for (var i = 0, l = urlParams.length; i < l; ++i) {
      var arg = args[0];
      if (urlParams[i] && !arg) {
        throw new Error('Shopify: method requires argument "' + urlParams[i] + '", but got: ' + arg);
      }
      urlData[urlParams[i]] = args.shift();
    }

    var requestPath = this.createFullPath(methodPath, urlData);

    self._request(requestMethod, requestPath, data, function(err, response) {
      if (err) {
        deferred.reject(err);
      } else {
        deferred.resolve(response);
      }
    });

    return deferred.promise;
  };
};

lib/shopify.js
/**
 * node-shopify-api
 *
 * Copyright (c) 2014 Chris Talkington, contributors.
 * Licensed under the MIT license.
 * https://github.com/ctalkington/node-shopify-api/blob/master/LICENSE-MIT
 */
var resources = {
  asset: require('./resources/asset'),
  assetLegacy: require('./resources/assetLegacy'),
  theme: require('./resources/theme')
};

function Shopify(options) {
  if (!(this instanceof Shopify)) {
    return new Shopify(options);
  }

  options = options || {};

  this._api = {
    auth: options.auth || null,
    host: options.host || Shopify.DEFAULT_HOST,
    port: options.port || Shopify.DEFAULT_PORT,
    timeout: options.timeout || Shopify.DEFAULT_TIMEOUT
  };

  this._initResources();
}

Shopify.DEFAULT_HOST = null;
Shopify.DEFAULT_PORT = 443;
Shopify.DEFAULT_TIMEOUT = 120000;

Shopify.prototype._initResources = function() {
  for (var name in resources) {
    this[name] = new resources[name](this);
  }
};

Shopify.prototype._setApiField = function(key, value) {
  this._api[key] = value;
};

Shopify.prototype.getApiField = function(key) {
  return this._api[key];
};

Shopify.prototype.setAuth = function(key, passwd) {
  if (key && passwd) {
    this._setApiField('auth', key + ':' + passwd);
  }
};

Shopify.prototype.setHost = function(host, port) {
  this._setApiField('host', host);
  if (port) {
    this.setPort(port);
  }
};

Shopify.prototype.setPort = function(port) {
  this._setApiField('port', port);
};

Shopify.prototype.setTimeout = function(timeout) {
  this._setApiField(
    'timeout',
    timeout === null ? Shopify.DEFAULT_TIMEOUT : timeout
  );
};

module.exports = Shopify;

lib/util/index.js
/**
 * node-shopify-api
 *
 * Copyright (c) 2014 Chris Talkington, contributors.
 * Licensed under the MIT license.
 * https://github.com/ctalkington/node-shopify-api/blob/master/LICENSE-MIT
 */
var util = module.exports = {};
util.extend = require('class-extend').extend;

util.isFunction = function(fn) {
  return typeof fn === 'function';
};

util.isObject = function(o) {
  return toString.call(o) === '[object Object]';
};

util.isString = function(str) {
  return typeof str === 'string';
};

// https://gist.github.com/padolsey/6008842
util.makeInterpolator = (function() {
  var rc = {
    '\n': '\\n', '\"': '\\\"',
    '\u2028': '\\u2028', '\u2029': '\\u2029'
  };
  return function makeInterpolator(str) {
    return new Function(
      'o',
      'return "' + (
        str
        .replace(/["\n\r\u2028\u2029]/g, function($0) {
          return rc[$0];
        })
        .replace(/\{([\s\S]+?)\}/g, '" + o["$1"] + "')
      ) + '";'
    );
  };
}());

util.stringifyRequestData = function(data) {
  if (typeof data === 'object' && Object.keys(data).length === 0) {
    data = '';
  } else {
    data = JSON.stringify(data);
  }

  return data;
};

lib/resources/asset.js
/**
 * node-shopify-api
 *
 * Copyright (c) 2014 Chris Talkington, contributors.
 * Licensed under the MIT license.
 * https://github.com/ctalkington/node-shopify-api/blob/master/LICENSE-MIT
 */
var Resource = require('../resource');

module.exports = Resource.extend({
  path: '/admin/themes/',

  create: Resource.method({
    method: 'PUT',
    path: '/{themeid}/assets.json',
    urlParams: ['themeid']
  }),

  destroy: Resource.method({
    method: 'DELETE',
    path: '/{themeid}/assets.json?asset[key]={key}',
    urlParams: ['themeid', 'key']
  }),

  list: Resource.method({
    method: 'GET',
    path: '/{themeid}/assets.json',
    urlParams: ['themeid']
  }),

  retrieve: Resource.method({
    method: 'GET',
    path: '/{themeid}/assets.json?asset[key]={key}&theme_id={themeid}',
    urlParams: ['themeid', 'key']
  }),

  update: Resource.method({
    method: 'PUT',
    path: '/{themeid}/assets.json',
    urlParams: ['themeid']
  }),
});

lib/resource.js
/**
 * node-shopify-api
 *
 * Copyright (c) 2014 Chris Talkington, contributors.
 * Licensed under the MIT license.
 * https://github.com/ctalkington/node-shopify-api/blob/master/LICENSE-MIT
 */
var https = require('https');
var path = require('path');

var when = require('when');

var SError = require('./error');
var util = require('./util');

function Resource(shopify) {
  this._shopify = shopify;
  this.path = util.makeInterpolator(this.path);

  this.initialize.apply(this, arguments);
}

Resource.extend = util.extend;
Resource.method = require('./method');

Resource.prototype.path = '';

Resource.prototype.initialize = function() {};

Resource.prototype.createFullPath = function(methodPath, data) {
  return path.join(
    this.path(data),
    typeof methodPath === 'function' ?
      methodPath(data) : methodPath
  ).replace(/\\/g, '/');
};

Resource.prototype.createDeferred = function(callback) {
  var deferred = when.defer();

  if (callback) {
    deferred.promise.then(function(res) {
      setTimeout(function(){ callback(null, res); }, 0);
    }, function(err) {
      setTimeout(function(){ callback(err, null); }, 0);
    });
  }

  return deferred;
};

Resource.prototype.createUrlData = function() {
  return {};
};

Resource.prototype._errorHandler = function(req, callback) {
  var self = this;

  return function(err) {
    if (req._isAborted) {
      return;
    }

    callback.call(self, new SError.ShopifyConnectionError({
      message: 'An error occurred with our connection to Shopify',
      detail: err
    }), null);
  };
};

Resource.prototype._parseCallLimit = function(headers) {
  var parsed = null;
  var data = {
    called: 0,
    limit: 0
  };

  if (headers && typeof headers['x-shopify-shop-api-call-limit'] === 'string') {
    parsed = headers['x-shopify-shop-api-call-limit'].split('/');
  }

  if (Array.isArray(parsed) && parsed.length === 2) {
    data.called = parseInt(parsed[0], 10);
    data.limit = parseInt(parsed[1], 10);
  }

  return data;
};

Resource.prototype._responseHandler = function(req, callback) {
  var self = this;

  return function(res) {
    var response = '';

    res.setEncoding('utf8');

    res.on('data', function(chunk) {
      response += chunk;
    });

    res.on('end', function() {
      var err = null;
      var resp = null;

      if (res.statusCode === 401) {
        return callback.call(self, new SError.ShopifyAuthenticationError({
          message: 'Authentication failed against the Shopify API'
        }), null);
      } else if (res.statusCode === 404) {
        return callback.call(self, new SError.ShopifyAPIError({
          message: 'Request failed with "HTTP 404 Not Found" from the Shopify API'
        }), null);
      } else if (res.statusCode === 429) {
        return callback.call(self, new SError.ShopifyCallLimitError({
          message: 'Request limit exceeded for the Shopify API',
          detail: self._parseCallLimit(res.headers)
        }), null);
      }

      try {
        resp = JSON.parse(response);

        if (resp.errors) {
          err = new SError.ShopifyInvalidRequestError({
            message: 'Errors returned from the Shopify API',
            detail: resp.errors
          });
        }
      } catch (e) {
        return callback.call(self, new SError.ShopifyAPIError({
          message: 'Invalid JSON received from the Shopify API',
          detail: e
        }), null);
      }

      callback.call(self, err, resp);
    });
  };
};

Resource.prototype._request = function(method, requestPath, data, callback) {
  var requestData = util.stringifyRequestData(data || {});
  var self = this;

  var headers = {
    'Accept': 'application/json',
    'Content-Type': 'application/json',
    'Content-Length': Buffer.byteLength(requestData)
  };

  function makeRequest() {
    var timeout = self._shopify.getApiField('timeout');

    var req = https.request({
      host: self._shopify.getApiField('host'),
      port: self._shopify.getApiField('port'),
      auth: self._shopify.getApiField('auth'),
      path: requestPath,
      method: method,
      headers: headers
    });

    req.setTimeout(timeout, self._timeoutHandler(timeout, req, callback));
    req.on('response', self._responseHandler(req, callback));
    req.on('error', self._errorHandler(req, callback));

    req.write(requestData);
    req.end();
  }

  makeRequest();
};

Resource.prototype._timeoutHandler = function(timeout, req, callback) {
  var self = this;

  return function() {
    var timeoutErr = new Error('ETIMEDOUT');
    timeoutErr.code = 'ETIMEDOUT';

    req._isAborted = true;
    req.abort();

    callback.call(self, new SError.ShopifyConnectionError({
      message: 'Request aborted due to timeout being reached (' + timeout + 'ms)',
      detail: timeoutErr
    }), null);
  };
};

module.exports = Resource;